File: apps/api/internal/modules/v1/checkout_session/create_checkout_session.go
package checkout_session

import (
	"errors"

	"github.com/useportcall/portcall/libs/go/apix"
	"github.com/useportcall/portcall/libs/go/routerx"
	cs "github.com/useportcall/portcall/libs/go/services/checkout_session"
)

// CreateCheckoutSessionRequest is the HTTP request body.
type CreateCheckoutSessionRequest struct {
	PlanID      string `json:"plan_id"`
	UserID      string `json:"user_id"`
	CancelURL   string `json:"cancel_url"`
	RedirectURL string `json:"redirect_url"`
}

// CreateCheckoutSession handles POST /v1/checkout-sessions.
func CreateCheckoutSession(c *routerx.Context) {
	var body CreateCheckoutSessionRequest
	if err := c.ShouldBindJSON(&body); err != nil {
		c.BadRequest("invalid request body")
		return
	}

	svc := cs.NewService(c.DB(), c.Crypto())
	result, err := svc.Create(&cs.CreateInput{
		AppID:       c.AppID(),
		PlanID:      body.PlanID,
		UserID:      body.UserID,
		CancelURL:   body.CancelURL,
		RedirectURL: body.RedirectURL,
	})
	if err != nil {
		var ve *cs.ValidationError
		if errors.As(err, &ve) {
			c.BadRequest(ve.Message)
			return
		}
		c.ServerError("error creating checkout session", err)
		return
	}

	response := new(apix.CheckoutSession).Set(result.Session)
	if result.CheckoutURL != "" {
		response.URL = result.CheckoutURL
	}
	c.OK(response)
}

File: libs/ui/src/api/create-checkout-session.ts
"use server";

import { cookies } from "next/headers";
import { redirect } from "next/navigation";
import { getPortcallClient } from "../lib/portcall";

export async function createCheckoutSession(planId: string) {
  const cookieStore = await cookies();

  const userId = cookieStore.get("user_id")?.value;

  if (!userId) {
    redirect("/login");
  }

  try {
    const portcall = getPortcallClient();
    const session = await portcall.checkoutSessions.create({
      plan_id: planId,
      user_id: userId,
      cancel_url: "http://localhost:3000?status=canceled", // TODO: fix
      redirect_url: "http://localhost:3000?status=success", // TODO: fix
    });

    return { ok: true, url: session.url };
  } catch (error) {
    console.error("Failed to create checkout session:", error);
    return { ok: false, message: "Failed to create checkout session" };
  }
}
